README.md
[![Build status](https://ci.appveyor.com/api/projects/status/e7c4l5272uwh0nic?svg=true)](https://ci.appveyor.com/project/devblackops/ovf-active-directory)

# OVF.Active.Directory

## Overview
This is a set of [Pester](https://github.com/pester/Pester) / [poshspec](https://github.com/Ticketmaster/poshspec) tests designed to test the
basic operation of an Active Directory domain controller. These Pester tests have been packaged into a module according to the
[Operation Validation Framework](https://github.com/PowerShell/Operation-Validation-Framework) layout.

### Current tests

* DNS
  * DNS Service status
  * DNSCache service status
  * TCP port 53 responds

* Global Catalog
  * TCP port 3268 responds

* AD Domain Services
  * NDTS service status
  * NetLogon service status

* Key Distribution Center
  * Kdc service status

* AD Web Services
  * ADWS service status
  * TCP port 9389 responds

* LDAP
  * TCP port 389 responds

* DFS
  * Dfs service status
  * DFSR service status

## Example Output
![Example Pester output](/Media/example.png)

CHANGELOG.md

## 1.0.0 (2016-08-14)
  * Initial release


Tests/Manifest.tests.ps1
$moduleName = $env:BHProjectName
$here = Split-Path -Parent $MyInvocation.MyCommand.Path
$root = Split-Path -Path $here -Parent
$modulePath = Join-Path -Path $root -ChildPath $moduleName
$manifestPath = Join-Path -Path $modulePath -Child "$moduleName.psd1"
$changelogPath = Join-Path -Path $root -Child "CHANGELOG.md"

Describe 'Module manifest' {
    Context 'Validation' {

        $script:manifest = $null

        It "has a valid manifest" {
            {
                $script:manifest = Test-ModuleManifest -Path $manifestPath -ErrorAction Stop -WarningAction SilentlyContinue
            } | Should Not Throw
        }

        It "has a valid name in the manifest" {
            $script:manifest.Name | Should Be $env:BHProjectName
        }        

        It "has a valid version in the manifest" {
            $script:manifest.Version -as [Version] | Should Not BeNullOrEmpty
        }
    
        It 'has a valid description' {
            $script:manifest.Description | Should Not BeNullOrEmpty
        }

        It 'has a valid author' {
            $script:manifest.Author | Should Not BeNullOrEmpty
        }
    
        It 'has a valid guid' {
            { 
                [guid]::Parse($script:manifest.Guid) 
            } | Should Not throw
        }
    
        It 'has a valid copyright' {
            $script:manifest.CopyRight | Should Not BeNullOrEmpty
        }
        
        # Only for DSC modules
        # It 'exports DSC resources' {
        #     $dscResources = ($Manifest.psobject.Properties | Where Name -eq 'ExportedDscResources').Value
        #     @($dscResources).Count | Should Not Be 0
        # }

        $script:changelogVersion = $null
        It "has a valid version in the changelog" {
            foreach ($line in (Get-Content $changelogPath)) {
                if ($line -match "^\D*(?<Version>(\d+\.){1,3}\d+)") {
                    $script:changelogVersion = $matches.Version
                    break
                }
            }
            $script:changelogVersion                | Should Not BeNullOrEmpty
            $script:changelogVersion -as [Version]  | Should Not BeNullOrEmpty
        }

        It "changelog and manifest versions are the same" {
            $script:changelogVersion -as [Version] | Should be ( $script:manifest.Version -as [Version] )
        }

        if (Get-Command git.exe -ErrorAction SilentlyContinue) {
            $script:tagVersion = $null
            It "is tagged with a valid version" -skip {
                $thisCommit = git.exe log --decorate --oneline HEAD~1..HEAD

                if ($thisCommit -match 'tag:\s*(\d+(?:\.\d+)*)') {
                    $script:tagVersion = $matches[1]
                }

                $script:tagVersion                  | Should Not BeNullOrEmpty
                $script:tagVersion -as [Version]    | Should Not BeNullOrEmpty
            }

            It "all versions are the same" {
                $script:changelogVersion -as [Version] | Should be ( $script:manifest.Version -as [Version] )
                #$script:manifest.Version -as [Version] | Should be ( $script:tagVersion -as [Version] )
            }
        }
    }
}


build.ps1
[cmdletbinding()]
param(
    [string[]]$Task = 'default'
)

function Resolve-Module {
    [Cmdletbinding()]
    param (
        [Parameter(Mandatory, ValueFromPipeline)]
        [string[]]$Name
    )

    Process {
        foreach ($ModuleName in $Name) {
            $Module = Get-Module -Name $ModuleName -ListAvailable
            Write-Verbose -Message "Resolving Module [$($ModuleName)]"
            
            if ($Module) {
                $Version = $Module | Measure-Object -Property Version -Maximum | Select-Object -ExpandProperty Maximum
                $GalleryVersion = Find-Module -Name $ModuleName -Repository PSGallery -Verbose:$false | 
                    Measure-Object -Property Version -Maximum | 
                    Select-Object -ExpandProperty Maximum

                if ($Version -lt $GalleryVersion) {                                        
                    Write-Verbose -Message "$($ModuleName) Installed Version [$($Version.tostring())] is outdated. Installing Gallery Version [$($GalleryVersion.tostring())]"
                    
                    Install-Module -Name $ModuleName -Verbose:$false -Force
                    Import-Module -Name $ModuleName -Verbose:$false -Force -RequiredVersion $GalleryVersion
                }
                else {
                    Write-Verbose -Message "Module Installed, Importing [$($ModuleName)]"
                    Import-Module -Name $ModuleName -Verbose:$false -Force -RequiredVersion $Version
                }
            }
            else {
                Write-Verbose -Message "[$($ModuleName)] Missing, installing Module"
                Install-Module -Name $ModuleName -Verbose:$false -Force
                Import-Module -Name $ModuleName -Verbose:$false -Force -RequiredVersion $Version
            }
        }
    }
}

Get-PackageProvider -Name Nuget -ForceBootstrap | Out-Null
Set-PSRepository -Name PSGallery -InstallationPolicy Trusted

'BuildHelpers', 'psake', 'poshspec' | Resolve-Module

Set-BuildEnvironment

Invoke-psake -buildFile "$PSScriptRoot\psake.ps1" -taskList $Task -nologo -Verbose:$VerbosePreference
exit ( [int]( -not $psake.build_success ) )



psake.ps1
# Define global variables used by tasks
properties {
    $projectRoot = $ENV:BHProjectPath
    if(-not $projectRoot) {
        $projectRoot = $PSScriptRoot
    }
    $sut = $env:BHProjectName
    $tests = "$projectRoot\Tests"   
}

task Default -depends Test

Task Init {
    $lines
    Set-Location $ProjectRoot
    "Build System Details:"
    Get-Item ENV:BH*

    $modules = 'Pester', 'PSDeploy', 'PSScriptAnalyzer'
    Install-Module $modules -Confirm:$false
    Import-Module $modules -Verbose:$false -Force
}

# Test entire project
task Test -depends Init, Analyze, Pester  {
}

task Analyze -Depends Init {
    $saResults = Invoke-ScriptAnalyzer -Path $sut -Severity Error -Recurse -Verbose:$false
    if ($saResults) {
        $saResults | Format-Table
        Write-Error -Message 'One or more Script Analyzer errors/warnings where found. Build cannot continue!'
    }
}

task Pester -Depends Init {
    if(-not $ENV:BHProjectPath) {
        Set-BuildEnvironment -Path $PSScriptRoot\..
    }
    Remove-Module $ENV:BHProjectName -ErrorAction SilentlyContinue
    Import-Module (Join-Path $ENV:BHProjectPath $ENV:BHProjectName) -Force

    $testResults = Invoke-Pester -Path $tests -PassThru
    if ($testResults.FailedCount -gt 0) {
        $testResults | Format-List
        Write-Error -Message 'One or more Pester tests failed. Build cannot continue!'
    }
}

task Deploy -depends Test {
    # Gate deployment
    if(
        $ENV:BHBuildSystem -ne 'Unknown' -and
        $ENV:BHBranchName -eq "master" -and
        $ENV:BHCommitMessage -match '!deploy'
    ) {
        $params = @{
            Path = "$projectRoot\module.psdeploy.ps1"
            Force = $true
            Recurse = $false
        }

        Invoke-PSDeploy @Params
    } else {
        "Skipping deployment: To deploy, ensure that...`n" +
        "`t* You are in a known build system (Current: $ENV:BHBuildSystem)`n" +
        "`t* You are committing to the master branch (Current: $ENV:BHBranchName) `n" +
        "`t* Your commit message includes !deploy (Current: $ENV:BHCommitMessage)"
    }
}


OVF.Active.Directory/Diagnostics/Simple/active.directory.tests.ps1
#require -modules poshspec

Import-Module -Name poshspec -Verbose:$false -ErrorAction Stop

describe 'DNS' {
    context 'Availability' {
        service DNS status { should be running }
        service Dnscache status { should be running }
        TcpPort localhost 53 TcpTestSucceeded { Should Be $true }        
    }
}

describe 'Global Catalog' {
    context 'Availability' {
        TcpPort localhost 3268 TcpTestSucceeded { Should Be $true }
    }
}

describe 'Active Directory Domain Services' {
    service NTDS status { should be running }
    service Netlogon status { should be running }    
}

describe 'Key Distribution Center' {
    context 'Availability' {
        service Kdc status { should be running }
    }
}

describe 'Active Directory Web Services' {
    context 'Availbility' {
        service ADWS status { should be running }
        TcpPort localhost 9389 TcpTestSucceeded { Should Be $true }
    }
}

describe 'LDAP' {
    context 'Availbility' {
        TcpPort localhost 389 TcpTestSucceeded { Should Be $true } 
    }
}

describe 'DFS' {
    context 'Availability' {
        service Dfs status { should be running }
        service DFSR status { should be running }
    }
}
